File: typing-core/src/main/java/net/avdw/typing/Analyser.java
package net.avdw.typing;

import java.util.HashMap;
import java.util.Map;
import org.apache.commons.math3.stat.Frequency;
import org.apache.commons.math3.stat.descriptive.DescriptiveStatistics;

public class Analyser {

    String presented;
    String transcribed;

    final int CHAR_PER_WORD = 5;
    final DescriptiveStatistics avgDelta;
    final DescriptiveStatistics avgPress;
    final Map<Character, DescriptiveStatistics> avgKeyDelta;
    final Map<Character, DescriptiveStatistics> avgKeyPress;
    final Map<Character, Frequency> keyErrorFreq;
    final Map<Character, Long> pressedKeys;

    private long last;
    private int errorIdx;

    Analyser(String presented) {
        this.presented = presented;
        transcribed = "";

        avgDelta = new DescriptiveStatistics();
        avgPress = new DescriptiveStatistics();
        avgKeyDelta = new HashMap();
        avgKeyPress = new HashMap();
        keyErrorFreq = new HashMap();
        pressedKeys = new HashMap();

        errorIdx = -1;

    }

    void start() {
        last = System.currentTimeMillis();
    }

    void keyDown(char c) {
        if (!pressedKeys.isEmpty()) {
            keyUp(pressedKeys.keySet().iterator().next());
        }
        pressedKeys.put(c, System.currentTimeMillis());

        if (transcribed.length() == presented.length()) {
            throw new IllegalStateException("Analyser complete");
        }

        long timeToLast = System.currentTimeMillis() - last;

        if (presented.charAt(transcribed.length()) == c) {
            transcribed += c; //http://www.keybr.com/practice

            if (!avgKeyDelta.containsKey(c)) {
                avgKeyDelta.put(c, new DescriptiveStatistics());
            }

            avgDelta.addValue(timeToLast);
            avgKeyDelta.get(c).addValue(timeToLast);
            
            if (transcribed.length() == presented.length()) {
                pressedKeys.clear();
            }
        } else if (errorIdx != transcribed.length()) {
            if (!keyErrorFreq.containsKey(presented.charAt(transcribed.length()))) {
                keyErrorFreq.put(presented.charAt(transcribed.length()), new Frequency());
            }

            keyErrorFreq.get(presented.charAt(transcribed.length())).addValue(c);
            errorIdx = transcribed.length();
        }

        last = System.currentTimeMillis();
    }

    void keyUp(Character c) {
        if (!pressedKeys.containsKey(c)) {
            return;
        }
        
        long pressDelta = System.currentTimeMillis() - pressedKeys.remove(c);
        if (!avgKeyPress.containsKey(c)) {
            avgKeyPress.put(c, new DescriptiveStatistics());
        }
        
        avgPress.addValue(pressDelta);
        avgKeyPress.get(c).addValue(pressDelta);
    }

    double cps() {
        return 1000d / avgDelta.getMean();
    }

    double wpm() {
        return cps() * 60 / CHAR_PER_WORD;
    }

    double acc() {
        double totalErr = keyErrorFreq.entrySet().stream().mapToLong(entry -> entry.getValue().getSumFreq()).sum();
        return 1 - totalErr / transcribed.length();
    }
}

File: typing-core/src/main/javascript/Analyser.js
function Analyser(presented) {
    this.presented = presented;
    this.transcribed = "";
    this.avgDelta = new DescriptiveStatistics();
    this.avgPress = new DescriptiveStatistics();
    this.avgKeyDelta = new HashMap();
    this.avgKeyPress = new HashMap();
    this.keyErrorFreq = new HashMap();
    this.pressedKeys = new HashMap();
    this.errorIdx = -1;
    this.CHAR_PER_WORD = 5;
    this.ignoreList = [];
}

Analyser.prototype = {
    constructor: Analyser,
    start: function () {
        this.last = new Date().getTime();
    },
    ignore: function() {
        this.ignoreList = this.ignoreList.concat([].slice.call(arguments));
    },
    keyDown: function (c) {
        if (this.ignoreList.includes(c)) {
            return;
        }
        if (!this.pressedKeys.isEmpty()) {
            this.keyUp(this.pressedKeys.keySet().pop());
        }
        this.pressedKeys.put(c, new Date().getTime());
        if (this.transcribed.length === this.presented.length) {
            throw "Analyser complete";
        }

        var timeToLast = new Date().getTime() - this.last;
        if (this.presented.charAt(this.transcribed.length) === c) {
            this.transcribed += c; //http://www.keybr.com/practice

            if (!this.avgKeyDelta.containsKey(c)) {
                this.avgKeyDelta.put(c, new DescriptiveStatistics());
            }

            this.avgDelta.addValue(timeToLast);
            this.avgKeyDelta.get(c).addValue(timeToLast);
            if (this.transcribed.length === this.presented.length) {
                this.pressedKeys.clear();
            }
        } else if (this.errorIdx !== this.transcribed.length) {
            if (!this.keyErrorFreq.containsKey(this.presented.charAt(this.transcribed.length))) {
                this.keyErrorFreq.put(this.presented.charAt(this.transcribed.length), new Frequency());
            }

            this.keyErrorFreq.get(this.presented.charAt(this.transcribed.length)).addValue(c);
            this.errorIdx = this.transcribed.length;
        }

        this.last = new Date().getTime();
    },
    keyUp: function (c) {
        if (this.ignoreList.includes(c)) {
            return;
        }
        if (!this.pressedKeys.containsKey(c)) {
            return;
        }

        var pressDelta = new Date().getTime() - this.pressedKeys.remove(c);
        if (!this.avgKeyPress.containsKey(c)) {
            this.avgKeyPress.put(c, new DescriptiveStatistics());
        }

        this.avgPress.addValue(pressDelta);
        this.avgKeyPress.get(c).addValue(pressDelta);
    },
    cps: function () {
        return 1000 / this.avgDelta.getMean();
    },
    wpm: function () {
        return this.cps() * 60 / this.CHAR_PER_WORD;
    },
    acc: function () {
        var totalErr = this.keyErrorFreq.entrySet().map(function (entry) {
            return entry.value.getSumFreq();
        }).reduce(function (prevFreq, freq) {
            return prevFreq + freq;
        }, 0);
        console.log(totalErr);
        return 1 - (totalErr / this.transcribed.length);
    }
};
        